elastic/kibana
# Log deprecation warnings for plugins which will no longer be disable-able in 8.0

__NOTE: This PR was replaced by https://github.com/elastic/kibana/pull/112602__

Closes https://github.com/elastic/kibana/issues/110614

## Notes if you are pinged for a CODEOWNERS review
This PR moves forward with the plan outlined in https://github.com/elastic/kibana/issues/89584 to remove the ability for plugins to be disable-able by default in 8.0.

If you are tagged for a review here, it is for one of the following reasons:
- Your plugin currently has an `enabled` property in your config schema, so a deprecation has been added for it.
- Your plugin currently _implicitly_ has an `enabled` property but does not have a config schema, however you had indicated you need to preserve the ability to disable it (e.g. `vis_type_*` plugins). In this case an explicit config has been added for you.

The good news is that most likely you'll only need to review one or two lines of code 🙂 

**To test the deprecations**:
1. Start Kibana with a default/empty logging config, and `<my-plugin>.enabled: true`
2. You should see the deprecation logged to the console
3. Deprecation should also be logged if `<my-plugin>.enabled: false`

I also tried to find any documentation that referred to a plugin with an `.enabled` property. If the docs for your app have been changed, please give them a look. (Preview: https://kibana_111890.docs-preview.app.elstc.co/diff)

This change will be backported to 7.16. In a follow-up PR, we will implement the actual 8.0 breaking change by removing the deprecated config.

## Notes for Core Team
The approach here does the following:
1. Adds a deprecation from the config service for any plugin with an `.enabled` config that's **implicitly** added (i.e., plugins which currently have no config schema)
2. Adds a config schema to a few `vis_type_*` plugins which currently don't have one but will need to keep an `.enabled` config in 8.0
3. Manually add a deprecation warning to plugins which already have an `.enabled` config that's **explicitly** added in their config schema
    - To make this easier I've added a new `deprecate` and `deprecateFromRoot` to the `ConfigDeprecationFactory`
4. Updates docs for any references to deprecated `.enabled` properties.

I went with this approach because:
- It's simple and explicit
- We will already need to manually change every existing config schema with an `.enabled` property in 8.0, so this is just a precursor to that PR
    - Sure, this means 2 PRs where we ping a bunch of codeowners instead of just one, but it also gives these teams a heads up
- It doesn't require any special treatment for 3rd party plugins:
    - 3rd party plugins which already have a config schema & an `.enabled` property are unaffected
    - 1st party and 3rd party plugins which currently have **no** config schema are treated in the same way (deprecation is logged)
    - If any 1st or 3rd party plugin wants to support an `.enabled` property moving forward, they just add it to their config schema

However, I'm still open to feedback on alternatives.

## Release note

Deprecates the ability for most plugins to be disabled using the `{plugin_name}.enabled` config option. In 8.0, most Kibana plugins can no longer be disabled using this option. Plugin developers can still opt-in to this feature by explicitly adding an `.enabled` property to their config schema.

If you are currently using one of these options in your Kibana config, please remove it before upgrading to 8.0.
